storage/innobase/lock/lock0lock.cc

```diff
@@ -5487,9 +5487,9 @@ dberr_t lock_sys_t::cancel(trx_t *trx, lock_t *lock, bool check_victim)
   {
     if (!lock_sys.rd_lock_try())
     {
-      mysql_mutex_unlock(&lock_sys.wait_mutex);
+      lock_sys.wait_mutex_unlock();
       lock_sys.rd_lock(SRW_LOCK_CALL);
-      mysql_mutex_lock(&lock_sys.wait_mutex);
+      lock_sys.wait_mutex_lock();
```

Comment:
We are holding shared `lock_sys.latch` here. I see that `lock_sys_t::cancel()` is (among others) being executed as part of `lock_wait()` that ends in a lock wait timeout, deadlock, or is interrupted by client disconnect or a `KILL` statement. It is also executed as part of `lock_trx_handle_wait()` that is invoked by `row_search_mvcc()` in special cases (`SKIP LOCKED`, and handling `UPDATE` at `READ COMMITTED` or `READ UNCOMMITTED` isolation level).

Does your test workload include those? If not, I would rather not add a spinloop here.

(Side note: The `end_wait:` label in `lock_wait()` can be moved after the call; each `goto end_wait` guarantees `!trx->lock.wait_lock`.)